<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [2]:
import pandas as pd
import psycopg2

In [3]:
# вставьте сюда параметры подключения из юнита 1. Работа с базой данных из Python 
DBNAME = 'project_sql'
USER = 'skillfactory'
PASSWORD = 'cCkxxLVrDE8EbvjueeMedPKt'
HOST = '84.201.134.129'
PORT = 5432

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [4]:
# Считаем по количеству id вакансий из таблицы vacancies
query_3_1 = f'''select 
           COUNT(id) 
           from vacancies
           
        '''

In [5]:
# результат запроса
df = pd.read_sql_query(query_3_1, connection)
df

C:\Users\Nina\AppData\Local\Temp\ipykernel_50084\1114379196.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


count
0  49197

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [6]:
# Считаем по количеству id работодателей из таблицы employers
query_3_2 = f'''select 
           COUNT(id)
           from employers
          
        '''

In [7]:
# результат запроса
df = pd.read_sql_query(query_3_2, connection)
df

C:\Users\Nina\AppData\Local\Temp\ipykernel_50084\3062646138.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_2, connection)


count
0  23501

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [8]:
# Считаем по количеству id регионов из таблицы areas
query_3_3 = f'''select 
           COUNT(id)
           from areas
          
        '''

In [9]:
# результат запроса
df = pd.read_sql_query(query_3_3, connection)
df

C:\Users\Nina\AppData\Local\Temp\ipykernel_50084\928149064.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_3, connection)


count
0   1362

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [11]:
# Считаем по количеству id сфер деятельности из таблицы industries
query_3_4 = f'''select 
           COUNT(id)
           from industries
          
        '''

In [12]:
# результат запроса
df = pd.read_sql_query(query_3_4, connection)
df

C:\Users\Nina\AppData\Local\Temp\ipykernel_50084\2101691108.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_4, connection)


count
0    294

***

# выводы по предварительному анализу данных
Вакансий (49197) практически в 2 раза больше, чем работодателей (23501). Значит, работодатели обычно размещают больше, чем одну вакансию.
Сфер деятельности, в которых ищут работников, 294, достаточно большое разнообразие. И, вероятно, что в каждой сфере деятельности разнообразный выбор вакансий.
Т.к. всего вакансий 49197, тогда на одну сферу приходится в среднем 167,33 вакансии.
Также достаточно разнообразный разброс по регионам поиска, а именно 1362. Но скорее всего количество вакансий по регионам будет не равномерным.
Неравномерость, скорее всего, связана с неравномерностью деловой активности по регионам России.

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [13]:
# Группироуем данные по названию региона, в каждом регионе считаем количество id вакансий
query_4_1 = f'''select 
           a.name, 
           COUNT(v.id)
           from vacancies v
           LEFT JOIN areas a on v.area_id=a.id
           GROUP BY a.name 
           ORDER BY COUNT(v.id) DESC
           LIMIT 5
          
        '''

In [14]:
# результат запроса
df = pd.read_sql_query(query_4_1, connection)
df

C:\Users\Nina\AppData\Local\Temp\ipykernel_50084\2457173557.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_1, connection)


name  count
0           Москва   5333
1  Санкт-Петербург   2851
2            Минск   2112
3      Новосибирск   2006
4           Алматы   1892

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [15]:
# Для отбора применяем WHERE, просим вывести количество id вакансий, где не нулевое значение полей заработной платы
query_4_2 = f'''select 
           COUNT(id)
           from vacancies 
           WHERE salary_from is not NULL or salary_to is not NULL
          
        '''

In [16]:
# результат запроса
df = pd.read_sql_query(query_4_2, connection)
df

C:\Users\Nina\AppData\Local\Temp\ipykernel_50084\970655580.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_2, connection)


count
0  24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [17]:
# Примеряем агрегатную функцию AVG, которая возвращает среднее значение
query_4_3 = f'''select 
           ROUND(AVG(salary_from)),
           ROUND(AVG(salary_to))
           from vacancies 
                    
        '''

In [18]:
# результат запроса
df = pd.read_sql_query(query_4_3, connection)
df

C:\Users\Nina\AppData\Local\Temp\ipykernel_50084\2500607024.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_3, connection)


round     round
0  71065.0  110537.0

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [19]:
# Группируем данные по типу рабочего графица и типу трудоустройства, считаем id вакансий
query_4_4 = f'''select 
           schedule,
           employment,
           COUNT(id)
           from vacancies 
           GROUP BY schedule, employment
           ORDER BY COUNT(id) DESC
           Limit 2
                    
        '''

In [20]:
# результат запроса
df = pd.read_sql_query(query_4_4, connection)
df

C:\Users\Nina\AppData\Local\Temp\ipykernel_50084\2022935265.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_4, connection)


schedule        employment  count
0       Полный день  Полная занятость  35367
1  Удаленная работа  Полная занятость   7802

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [21]:
# Группируем данные по опыту работы, считаем количество id вакансий в сформированных группах. С помощью ORDER BY сортируем 
query_4_5 = f'''select 
           experience,
           COUNT(id)
           from vacancies 
           GROUP BY 1
           ORDER BY COUNT(id) 
                               
        '''

In [22]:
# результат запроса
df = pd.read_sql_query(query_4_5, connection)
df

C:\Users\Nina\AppData\Local\Temp\ipykernel_50084\2316260449.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_5, connection)


experience  count
0         Более 6 лет   1337
1           Нет опыта   7197
2       От 3 до 6 лет  14511
3  От 1 года до 3 лет  26152

***

# выводы по детальному анализу вакансий
Больше всего на hh работодатели ищут молодых специалистов с опытом работы от 1 до 3 лет, на полный день и полную занятость.
Средняя зарплатная вилка 70-110 тыс. руб. В половине случаев работодатели в вакансии не указывают зарплатную вилку.
Основные регионы по наличию вакансий: Москва, Санкт-Петербург, Минск, Новосибирск, Алматы

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [23]:
# Группируем данные по наименованию работодателя, считаем количество вакансий по id
query_5_1 = f'''select 
           e.name,
           COUNT(v.id)
           from employers e
           left JOIN vacancies v on e.id = v.employer_id
           GROUP BY e.name
           ORDER BY COUNT(v.id) DESC
           Limit 5
           
                               
        '''

In [24]:
# результат запроса
df = pd.read_sql_query(query_5_1, connection)
df

C:\Users\Nina\AppData\Local\Temp\ipykernel_50084\3834399874.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_1, connection)


name  count
0         Яндекс   1933
1     Ростелеком    491
2       Тинькофф    444
3           СБЕР    428
4  Газпром нефть    331

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [11]:
# Группируем данные по названию региона, считаем количество вакансий и работодателей в регионах
query_5_2 = f'''select 
          a.name area, 
          count(DISTINCT v.id) vacation, 
          count(distinct e.id) company
          from areas a
          left join vacancies v on a.id = v.area_id
          left join employers e on a.id = e.area
          where v.area_id is null
          group by a.name
          order by  3 desc
          limit 1
           
                               
        '''


In [12]:
# результат запроса
df = pd.read_sql_query(query_5_2, connection)
df

C:\Users\Nina\AppData\Local\Temp\ipykernel_48260\2250345976.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_2, connection)


area  vacation  company
0  Россия         0      410

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [25]:
# Группируем данные по названию региона, считаем число регионов, сортируем
query_5_3 = f'''select 
          e.name , 
          count(DISTINCT v.area_id) "Число регионов"
          from employers e
          join vacancies v on e.id = v.employer_id
          group by e.name
          order by 2 desc
          limit 1
                            
        '''

In [26]:
# результат запроса
df = pd.read_sql_query(query_5_3, connection)
df

C:\Users\Nina\AppData\Local\Temp\ipykernel_50084\201667002.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_3, connection)


name  Число регионов
0  Яндекс             181

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [27]:
# Применяем HAVING для отбора id работодателей, у которых не указана сфера деятельности
query_5_4 = f'''select 
          count(DISTINCT e.id)
          from employers e
          left join employers_industries i on e.id = i.employer_id
          GROUP BY i.industry_id
          HAVING i.industry_id is NULL
                                              
        '''

In [28]:
# результат запроса
df = pd.read_sql_query(query_5_4, connection)
df

C:\Users\Nina\AppData\Local\Temp\ipykernel_50084\3493105328.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_4, connection)


count
0   8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [29]:
# Группируем данные на названию компании, отбираем с помощью HAVING
query_5_5 = f'''select 
          e.name,
          count(DISTINCT i.industry_id)
          from employers e
          left join employers_industries i on e.id = i.employer_id
          GROUP BY e.name
          HAVING count(DISTINCT i.industry_id) =4
          ORDER BY e.name
          Limit 3
                                              
        '''

In [30]:
# результат запроса
df = pd.read_sql_query(query_5_5, connection)
df

C:\Users\Nina\AppData\Local\Temp\ipykernel_50084\388711819.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_5, connection)


name  count
0  101 Интернет      4
1      21vek.by      4
2          2ГИС      4

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [31]:
# Отбираем работодателей из сферы разработки ПО с помощью HAVING, считаем количество
query_5_6 = f'''select 
          count (distinct e.id) employer
          from employers_industries i
          left join employers e on e.id = i.employer_id
          left join industries ind on ind.id = i.industry_id
          WHERE ind.name = 'Разработка программного обеспечения'
        
                                                            
        '''

In [32]:
# результат запроса
df = pd.read_sql_query(query_5_6, connection)
df

C:\Users\Nina\AppData\Local\Temp\ipykernel_50084\821628714.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_6, connection)


employer
0      3553

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [33]:
# код для получения списка городов-милионников
import requests # Импортируем библиотеку requests
from bs4 import BeautifulSoup # Импортируем библиотеку BeautifulSoup
url = 'https://ru.wikipedia.org/wiki/Города-миллионеры_России' # Определяем значение URL страницы для запроса
response = requests.get(url) # Делаем GET-запрос к ресурсу и результат ответа сохраняем в переменной response
page = BeautifulSoup(response.text, 'html.parser') # Создаём объект BeautifulSoup, указывая html-парсер
tabl = page.find("table", class_="standard")       # ищем на странице таблицу

town = []
tt = tabl.find("tbody") # в таблице выбираем тело таблицы
qq = tt.find_all("tr")  # в теле таблицы выбираем все ряды

for ss in qq:
  qqq = list(ss.find("a"))  # в каждом ряде находим тег "а", преобразуем его в список
  town.append(qqq[0])       # и добавляем содержимое к списку городов
town = town[1:]             # в первом ряду кусок от названия, отсекаем его

In [34]:
# текст запроса
query = f'''
          (select a.name area, count(v.id) count -- count(DISTINCT ei.industry_id) industries--, count(distinct e.id) company
          from employers e
          join vacancies v on e.id = v.employer_id
          join areas a on v.area_id = a.id
          where (e.name = 'Яндекс')'''
# далее формируем строку для фильтрации городов миллионников из предыдущего блока
# в результате которого получили список городов миллионников
ll = len(town)
if ll > 0:
  str = f" AND (( a.name = '{town[0]}')" # если есть хоть один город миллионник
  if ll > 1:
    for s in town[1:]:
      str += f" OR (a.name = '{s}')"     # и для каждого следующего города из списка

# далее завершение запроса
query += (str + f''')
          group by a.name
          order by  2 desc)
          '''
          +                             # далее суммирем результат
'''
          --limit 6
          union  all
          select 'Total :' area, count( v.id)
          from employers e
          join vacancies v on e.id = v.employer_id
          join areas a on v.area_id = a.id
          where (e.name = 'Яндекс')'''
          +str
          +
'''       )
          group by area
          order by  2
'''
)

In [35]:
# результат запроса
df = pd.read_sql_query(query, connection)
df

C:\Users\Nina\AppData\Local\Temp\ipykernel_50084\1115395102.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


area  count
0              Омск     21
1         Челябинск     22
2        Красноярск     23
3         Волгоград     24
4            Казань     25
5    Ростов-на-Дону     25
6             Пермь     25
7            Самара     26
8               Уфа     26
9         Краснодар     30
10          Воронеж     32
11      Новосибирск     35
12  Нижний Новгород     36
13     Екатеринбург     39
14  Санкт-Петербург     42
15           Москва     54
16          Total :    485

***

Больше всего вакансий размещает компания Яндекс	(1933), далее в 3,9 раз меньше размещает вакансий компания Ростелеком, затем Тинькофф, СБЕР и Газпромнефть.
Технологичные компании размещают больше всего вакансий.
Яндекс размещает вакансии в самом большом количестве разных регионов.
3553 работодателей ищут сотрудников в сфере Разработка ПО (достаточно большой выбор работодателей).


# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [4]:
# Отбираем с помощью WHERE
query_6_1 = f'''select 
          count(v.id)
          from vacancies v
          where  (v.name iLIKE '%data%') OR  (v.name iLIKE '%данн%')
          order by  1
          
        '''

In [5]:
# результат запроса
df = pd.read_sql_query(query_6_1, connection)
df

C:\Users\Nina\AppData\Local\Temp\ipykernel_26484\366702258.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_1, connection)


count
0   1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [8]:
# Отбираем с помощью Where и указанием слов, соответствующих условию
query_6_2 = f'''select
    count(v.name)
    from vacancies v
    WHERE
(
    (name ILIKE '%data scientist%') OR (name ILIKE '%data science%') OR (name ILIKE '%исследователь данных%')
    OR ((name ILIKE '%ML%') AND (name NOT ILIKE '%HTML%'))
    OR (name ILIKE '%machine learning%')
    OR (name ILIKE '%машинн%обучен%')
    AND (key_skills IS NOT NULL)
)

    AND (v.name ILIKE '%junior%'
    OR v.experience ILIKE '%Нет опыта%'
    OR v.employment ILIKE '%Стажировка%')
    '''

In [9]:
# результат запроса
df = pd.read_sql_query(query_6_2, connection)
df

C:\Users\Nina\AppData\Local\Temp\ipykernel_26484\2267038862.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_2, connection)


count
0     51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [19]:
# Добавляем в WHERE условия по ключевым навыкам
query_6_3 = f'''select 
     COUNT(v.id)

from vacancies v

WHERE ( (v.key_skills like '%SQL%')
        or
        (v.key_skills ilike '%postgres%')
        )
    AND
(
    (name ILIKE '%data scientist%') OR (name ILIKE '%data science%') OR (name ILIKE '%исследователь данных%')
    OR ((name ILIKE '%ML%') AND (name NOT ILIKE '%HTML%'))
    OR (name ILIKE '%machine learning%')
    OR (name ILIKE '%машинн%обучен%')
    AND (key_skills IS NOT NULL)
)
    

   '''

In [20]:
# результат запроса
df = pd.read_sql_query(query_6_3, connection)
df

C:\Users\Nina\AppData\Local\Temp\ipykernel_26484\3794299702.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_3, connection)


count
0    229

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [25]:
# Добавляем в WHERE условие отбора - ключевой навык Python
query_6_4 = f'''select 
     COUNT(v.id)

from vacancies v

WHERE v.key_skills ilike '%Python%'
        
    AND
(
    (name ILIKE '%data scientist%') OR (name ILIKE '%data science%') OR (name ILIKE '%исследователь данных%')
    OR ((name ILIKE '%ML%') AND (name NOT ILIKE '%HTML%'))
    OR (name ILIKE '%machine learning%')
    OR (name ILIKE '%машинн%обучен%')
    AND (key_skills IS NOT NULL)
)
    

   '''

In [26]:
# результат запроса
df = pd.read_sql_query(query_6_4, connection)
df

C:\Users\Nina\AppData\Local\Temp\ipykernel_26484\4264665179.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_4, connection)


count
0    357

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [27]:
# Отбираем вакансии для DS, вычисляем среднее значение
query_6_5 = f'''SELECT 
                    ROUND(AVG(
                        LENGTH(v.key_skills) - LENGTH(REPLACE(v.key_skills, CHR(9), '')) + 1
                        ), 2) AS "Среднее количество навыков"
                FROM public.vacancies v
                WHERE 
                    (v.name ILIKE '%data scientist%' OR 
                    v.name ILIKE '%data science%' OR
                    v.name ILIKE '%исследователь данных%' OR
                    (v.name LIKE '%ML%' AND v.name NOT ILIKE '%HTML%') OR
                    v.name ILIKE '%machine learning%' OR
                    v.name ILIKE '%машинн%обучен%%') 
'''

In [28]:
# результат запроса
df = pd.read_sql_query(query_6_5, connection)
df

C:\Users\Nina\AppData\Local\Temp\ipykernel_26484\3654210517.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_5, connection)


Среднее количество навыков
0                        6.41

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [29]:
# Считаем среднее значение заработной платы для дата-саентистов с разным опытом работы
query_6_6 = f'''select

  v.experience, ROUND(AVG(coalesce((v.salary_from + v.salary_to) / 2, v.salary_from, v.salary_to)), 0)

from vacancies v

WHERE
(
       (v.name ILIKE '%data scientist%')
    OR (lower(v.name) ILIKE '%data science%')
    OR (v.name ILIKE '%исследователь данных%')
    OR (v.name  LIKE '%ML%' AND  (v.name NOT iLIKE '%html%'))
    OR (v.name ILIKE '%machine learning%')
    OR (v.name ILIKE '%машинн%обучен%')
    AND (key_skills IS NOT NULL)
)

GROUP BY v.experience
ORDER BY 1
'''

In [30]:
# результат запроса
df = pd.read_sql_query(query_6_6, connection)
df

C:\Users\Nina\AppData\Local\Temp\ipykernel_26484\519194770.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_6, connection)


experience     round
0         Более 6 лет       NaN
1           Нет опыта   74643.0
2  От 1 года до 3 лет  145236.0
3       От 3 до 6 лет  243115.0

***

Всего вакансий по работе с данными 1771, для начинающих DS (без опыта) 51, т.е. 2,8%.
В основном работодатели ищут специалистов с опытом работы.
В качестве ключевого опыта SQL указано 229 вакансий, для Pethon 357.
Целесообразно понять, какой еще опыт требуется для DS (т.к. всего в среднем работодатели указывают 6.41 ключевых навыков).
Каждые 3 года опыта увеличивают заработную плату в 2 раза.

# Общий вывод по проекту

Самое больше количество вакансий у технологичных компаний. Это отражает развитие мирового тренда цифровой экономики и экономики данных. 
Работодатели в сфере DS хотят нанимать работников с опытом работы.
Распределение вакансий по регионам не равномерно. Самое большое количество в городах миллионниках, столицах. 
Работников в основном ищут на полный день и полную занятость.

Можно провести исследование рынка удаленной работы (т.к. это даст возможность работать удаленно из регионов и закрыть вакансии в городах миллионниках). Провести исследование по годам, с какой скоростью растет запрос на технологичных работников и предсказать тренд - какая потребность будет в будущем. Какие навыки в будущем будут актуальны. Определить ключевые навыки - которые влекут резкое увеличение заработной платы.

Дополнительное исследование проводится с целью понять, какие навыки развивать с целью работы в Data Science.

Цель дополнительного исследования понять, какие еще навыки нужны для DS, помимо Python и SQL. Вывести ТОП-15.

In [105]:
# Запрос для понимания сколько всего вакансий размещено в DS
query_7_0 = f'''select
    count(v.id)
    from vacancies v
    WHERE
(
    (name ILIKE '%data scientist%') OR (name ILIKE '%data science%') OR (name ILIKE '%исследователь данных%')
    OR ((name ILIKE '%ML%') AND (name NOT ILIKE '%HTML%'))
    OR (name ILIKE '%machine learning%')
    OR (name ILIKE '%машинн%обучен%')
    AND (key_skills IS NOT NULL)
)
    '''

In [106]:
df = pd.read_sql_query(query_7_0, connection)
df

C:\Users\Nina\AppData\Local\Temp\ipykernel_9604\179480186.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_7_0, connection)


count
0    533

Всего для DS 533 вакансии, из них 51 вакансия для начинающих 9,5%.
Из 533 вакансий, в качестве ключевого опыта SQL указано 229 вакансий (42,9%), для Python 357 (66,97%).
Какие еще требуются навыки, как часто их указывают?

In [133]:
# Создаем таблицу со всеми ключевыми навыками, которые указывают в DS
query_7 = f'''SELECT 
                    v.key_skills
                    
                FROM public.vacancies v
                WHERE 
                   (name ILIKE '%data scientist%') OR (name ILIKE '%data science%') OR (name ILIKE '%исследователь данных%')
    OR ((name ILIKE '%ML%') AND (name NOT ILIKE '%HTML%'))
    OR (name ILIKE '%machine learning%')
    OR (name ILIKE '%машинн%обучен%')
    AND (key_skills IS NOT NULL)
'''

In [134]:
df = pd.read_sql_query(query_7, connection)
df

C:\Users\Nina\AppData\Local\Temp\ipykernel_9604\2767996417.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_7, connection)


key_skills
0            Python\tSQL\tРабота в команде\tOpenCV\tML
1    Oracle Pl/SQL\tБазы данных\tMachine Learning\t...
2    Python\tOpenCV\tTensorflow\tComputer Vision\tН...
3                       Python\tFlask\tLinux\tAWS\tGit
4    Atlassian Jira\tФункциональное тестирование\tS...
..                                                 ...
528  Qt\tООП\tC++\tMS Visual Studio\tРазработка ком...
529  SCALA\tSQL\tМатематический анализ\tСтатистичес...
530  Python\tLinux\tKubernetes\tAWS\tGit\tPeople Ma...
531  UML\tSQL\tBPMN\tSOAP\tREST\tПостановка задач р...
532  Git\tJavaScript\tHTML\tCSS\tVue.js\tVue\tREST ...

[533 rows x 1 columns]

In [135]:
# Создаем список ключевых навыков
skill_list = []
for object in df['key_skills']:
   if not object:
      continue
   else:
      myList = object.split('\t')
      for word in myList:
         if word not in skill_list:  # если строка не в списке
            skill_list.append(word)  # добавляем её

skill_list.append('Не заполнено')     
print(skill_list)
print(len(skill_list))
   

['Python', 'SQL', 'Работа в команде', 'OpenCV', 'ML', 'Oracle Pl/SQL', 'Базы данных', 'Machine Learning', 'Машинное обучение', 'Git', 'Data science', 'Работа с базами данных', 'Математические модели', 'Tensorflow', 'Computer Vision', 'Нейронные сети', 'C++', 'С#', 'Caffe', 'Flask', 'Linux', 'AWS', 'Atlassian Jira', 'Функциональное тестирование', 'Тестирование', 'Blockchain', 'Регресcионное тестирование', 'Jira', 'Автоматизированное тестирование', 'Java', 'Проведение тестирований', 'Нагрузочное тестирование', 'Разработка ПО', 'JavaScript', 'MySQL', 'Веб-программирование', 'CSS', 'PHP5', 'ООП', 'PHP', 'Постановка задач разработчикам', 'руководство командой разработчиков', 'HTML5', 'laravel', 'vue', 'react', 'DDD', 'devops', 'NLP', 'Scikit-learn', 'Поиск информации в интернет', 'Математическая статистика', 'Математический анализ', 'Qt', 'VoIP', 'TCP/IP', 'QML', 'Bash', 'Sas', 'MS PowerPoint', 'Анализ данных', 'Data Mining', 'machine learning', 'SCALA', 'Big Data', 'Data Analysis', 'MATLAB

Всего в вакансиях указывают 530 различных ключевых навыков.

In [136]:
# Из списка ключевых навыков делаем словарь, где ключевые навыки являются ключами
empty_dict = dict.fromkeys(skill_list)
print(empty_dict)

{'Python': None, 'SQL': None, 'Работа в команде': None, 'OpenCV': None, 'ML': None, 'Oracle Pl/SQL': None, 'Базы данных': None, 'Machine Learning': None, 'Машинное обучение': None, 'Git': None, 'Data science': None, 'Работа с базами данных': None, 'Математические модели': None, 'Tensorflow': None, 'Computer Vision': None, 'Нейронные сети': None, 'C++': None, 'С#': None, 'Caffe': None, 'Flask': None, 'Linux': None, 'AWS': None, 'Atlassian Jira': None, 'Функциональное тестирование': None, 'Тестирование': None, 'Blockchain': None, 'Регресcионное тестирование': None, 'Jira': None, 'Автоматизированное тестирование': None, 'Java': None, 'Проведение тестирований': None, 'Нагрузочное тестирование': None, 'Разработка ПО': None, 'JavaScript': None, 'MySQL': None, 'Веб-программирование': None, 'CSS': None, 'PHP5': None, 'ООП': None, 'PHP': None, 'Постановка задач разработчикам': None, 'руководство командой разработчиков': None, 'HTML5': None, 'laravel': None, 'vue': None, 'react': None, 'DDD': No

In [137]:
# В ранее созданном словаре были пустые значения, к которым нельзя было прибавлять единицу (сделать счетчик, чтоб считать сколько раз встречается каждый навык)
# Поэтому подготовительным этапом ко всем ключам добавили значение 0
skill_dict={}

for object in df['key_skills']:
   if not object:
      continue
   else:
      newList = object.split('\t')
      for skill in newList:
         if skill in skill_list:
            skill_dict[skill]=0
print(skill_dict)

{'Python': 0, 'SQL': 0, 'Работа в команде': 0, 'OpenCV': 0, 'ML': 0, 'Oracle Pl/SQL': 0, 'Базы данных': 0, 'Machine Learning': 0, 'Машинное обучение': 0, 'Git': 0, 'Data science': 0, 'Работа с базами данных': 0, 'Математические модели': 0, 'Tensorflow': 0, 'Computer Vision': 0, 'Нейронные сети': 0, 'C++': 0, 'С#': 0, 'Caffe': 0, 'Flask': 0, 'Linux': 0, 'AWS': 0, 'Atlassian Jira': 0, 'Функциональное тестирование': 0, 'Тестирование': 0, 'Blockchain': 0, 'Регресcионное тестирование': 0, 'Jira': 0, 'Автоматизированное тестирование': 0, 'Java': 0, 'Проведение тестирований': 0, 'Нагрузочное тестирование': 0, 'Разработка ПО': 0, 'JavaScript': 0, 'MySQL': 0, 'Веб-программирование': 0, 'CSS': 0, 'PHP5': 0, 'ООП': 0, 'PHP': 0, 'Постановка задач разработчикам': 0, 'руководство командой разработчиков': 0, 'HTML5': 0, 'laravel': 0, 'vue': 0, 'react': 0, 'DDD': 0, 'devops': 0, 'NLP': 0, 'Scikit-learn': 0, 'Поиск информации в интернет': 0, 'Математическая статистика': 0, 'Математический анализ': 0, '

In [138]:
# Посчитали сколько каждый навык встречается в ключевых навыках вакансий
for object in df['key_skills']:
   if not object:
      continue
   else:
      newList = object.split('\t')
      for skill in newList:
         if skill in skill_list:
            skill_dict[skill]+=1
print(skill_dict)

{'Python': 354, 'SQL': 208, 'Работа в команде': 10, 'OpenCV': 14, 'ML': 49, 'Oracle Pl/SQL': 9, 'Базы данных': 11, 'Machine Learning': 114, 'Машинное обучение': 31, 'Git': 81, 'Data science': 4, 'Работа с базами данных': 8, 'Математические модели': 1, 'Tensorflow': 35, 'Computer Vision': 19, 'Нейронные сети': 8, 'C++': 42, 'С#': 1, 'Caffe': 2, 'Flask': 4, 'Linux': 53, 'AWS': 5, 'Atlassian Jira': 14, 'Функциональное тестирование': 2, 'Тестирование': 2, 'Blockchain': 1, 'Регресcионное тестирование': 1, 'Jira': 2, 'Автоматизированное тестирование': 2, 'Java': 41, 'Проведение тестирований': 1, 'Нагрузочное тестирование': 1, 'Разработка ПО': 4, 'JavaScript': 13, 'MySQL': 9, 'Веб-программирование': 2, 'CSS': 3, 'PHP5': 1, 'ООП': 27, 'PHP': 10, 'Постановка задач разработчикам': 10, 'руководство командой разработчиков': 4, 'HTML5': 1, 'laravel': 1, 'vue': 1, 'react': 1, 'DDD': 1, 'devops': 1, 'NLP': 40, 'Scikit-learn': 14, 'Поиск информации в интернет': 1, 'Математическая статистика': 62, 'Мат

In [151]:
# Сделали из словаря таблицу, отсортировали по количеству указаний ключевого навыка, вывели ТОП-15
df_skills = pd.DataFrame.from_dict(skill_dict, orient='index')
df_sorted = df_skills.sort_values(by=0, ascending=False)
df_sorted.head(15)

0
Python                     354
SQL                        208
Machine Learning           114
Git                         81
Математическая статистика   62
Data Analysis               55
Linux                       53
Pandas                      52
Data Science                52
ML                          49
Английский язык             48
Анализ данных               46
PyTorch                     46
Big Data                    42
C++                         42

In [152]:
# Обрезали таблицу до ТОП 15, чтоб на ее основе построить диаграмму
cut_data = df_sorted[:15]
cut_data

0
Python                     354
SQL                        208
Machine Learning           114
Git                         81
Математическая статистика   62
Data Analysis               55
Linux                       53
Pandas                      52
Data Science                52
ML                          49
Английский язык             48
Анализ данных               46
PyTorch                     46
Big Data                    42
C++                         42

In [ ]:
# Строим для наглядности столбчатую диаграмму ТОП-15 ключевых навыков, которые указывают в вакансиях по DS
import plotly.express as px
fig = px.bar(cut_data, 
             x=0, 
             title='Топ-15 ключевых навыков для Data Science')
fig.show()

Вывели список ключивых навыков, которые необходимо развивать для развития в сфере Data Science

Новое изменение, создала ветвь master в папке Project_2